src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    !matchingItems.Any())
+                {
+                    // The user has typed something, but nothing in the actual list matched what
+                    // they were typing.  In this case, we want to dismiss completion entirely.
+                    // The thought process is as follows: we aggressively brought up completion
+                    // to help them when they typed delete (in case they wanted to pick another
+                    // item).  However, they're typing something that doesn't seem to match at all
+                    // The completion list is just distracting at this point.
+                    return null;
+                }
+
+                MatchResult<VSCompletionItem>? bestMatchResult = null;
+                var moreThanOneMatchWithSamePriority = false;
+                foreach (var currentMatchResult in matchingItems)
+                {
+                    if (bestMatchResult == null)
+                    {
+                        // We had no best result yet, so this is now our best result.
+                        bestMatchResult = currentMatchResult;
+                    }
+                    else
+                    {
+                        var match = currentMatchResult.CompareTo(bestMatchResult.Value, _filterText);
+                        if (match > 0)
+                        {
+                            moreThanOneMatchWithSamePriority = false;
+                            bestMatchResult = currentMatchResult;
+                        }
+                        else if (match == 0)
+                        {
+                            moreThanOneMatchWithSamePriority = true;
+                        }
+                    }
+                }
+
+                int index;
+                bool hardSelect;
+
+                // If we had a matching item, then pick the best of the matching items and
+                // choose that one to be hard selected.  If we had no actual matching items
+                // (which can happen if the user deletes down to a single character and we
+                // include everything), then we just soft select the first item.
+                if (bestMatchResult != null)
+                {
+                    // Only hard select this result if it's a prefix match
+                    // We need to do this so that
+                    // * deleting and retyping a dot in a member access does not change the
+                    //   text that originally appeared before the dot
+                    // * deleting through a word from the end keeps that word selected
+                    // This also preserves the behavior the VB had through Dev12.
+                    hardSelect = !_hasSuggestedItemOptions && bestMatchResult.Value.EditorCompletionItem.FilterText.StartsWith(_filterText, StringComparison.CurrentCultureIgnoreCase);
+                    index = items.IndexOf(bestMatchResult.Value);
```

Comment:
it feels like index can be set during the above loop instead of havin got find teh item again.  

i'm on the fence about hardSelect.  it does seem nice to only compute that once.